File: Server/logClient.py
#!/usr/bin/env python3
import paho.mqtt.client as mqtt
import sys

# Broker settings
BROKER_ADDRESS = "localhost"
CLIENT_ID = "AsyncSubscriber"
TOPIC = "test/topic"
QOS = 1

# Callback if the connection is lost or disconnected
def on_disconnect(client, userdata, rc):
    print("[ASYNC] Disconnected from broker with return code:", rc)

# Callback when the client receives a CONNACK response from the broker
def on_connect(client, userdata, flags, rc):
    if rc == 0:
        print(f"[ASYNC] Connected to broker at {BROKER_ADDRESS}")
        # Now subscribe to the topic
        client.subscribe(TOPIC, qos=QOS)
        print(f"[ASYNC] Subscribed to topic: {TOPIC}")
    else:
        print("Failed to connect, return code %d" % rc)

# Callback when a PUBLISH message is received from the broker
def on_message(client, userdata, msg):
    print(f"[ASYNC] Message arrived on topic '{msg.topic}': {msg.payload.decode()}")

# Optionally, callback when a message with QoS>0 is delivered
def on_publish(client, userdata, mid):
    print("[ASYNC] Delivery confirmed for mid:", mid)

def main():
    # Create a new MQTT client instance
    client = mqtt.Client(client_id=CLIENT_ID)

    # Set callbacks
    client.on_connect = on_connect
    client.on_message = on_message
    client.on_disconnect = on_disconnect
    client.on_publish = on_publish

    # Connect to the broker
    try:
        client.connect(BROKER_ADDRESS, 1883, 60)
    except Exception as e:
        print(f"Failed to connect: {e}")
        sys.exit(1)

    print("[ASYNC] Press Ctrl+C to quit...")

    # Start the network loop, which runs in the foreground here.
    # This call will block and automatically handle reconnections, pinging, etc.
    client.loop_forever()

if __name__ == "__main__":
    main()

File: Server/logClient.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "MQTTClient.h"

/* Define broker settings */
#define ADDRESS     "tcp://localhost:1883"
#define CLIENTID    "AsyncSubscriber"
#define TOPIC       "test/topic"
#define QOS         1

/* Callback if the connection is lost unexpectedly */
void connectionLost(void* context, char* cause)
{
    printf("[ASYNC] Connection lost: %s\n", cause);
}

/* Callback when a subscribed message arrives */
int messageArrived(void* context, char* topicName, int topicLen, MQTTClient_message* message)
{
    /* Print the payload as a string */
    printf("[ASYNC] Message arrived on topic '%s': %.*s\n",
           topicName,
           message->payloadlen,
           (char*)message->payload);

    /* Free the message and the topic name memory allocated by the library */
    MQTTClient_freeMessage(&message);
    MQTTClient_free(topicName);

    /* Return 1 to indicate that the message has been fully processed */
    return 1;
}

/* Callback when a message published with QoS>0 is delivered */
void deliveryComplete(void* context, MQTTClient_deliveryToken dt)
{
    printf("[ASYNC] Delivery confirmed for token %d\n", dt);
}

int main(int argc, char* argv[])
{
    MQTTClient client;
    MQTTClient_connectOptions conn_opts = MQTTClient_connectOptions_initializer;
    int rc;

    /* 1) Create the asynchronous client */
    rc = MQTTClient_create(&client, ADDRESS, CLIENTID,
                           MQTTCLIENT_PERSISTENCE_NONE, NULL);
    if (rc != MQTTCLIENT_SUCCESS)
    {
        printf("Failed to create client, return code %d\n", rc);
        return EXIT_FAILURE;
    }

    /* 2) Set callbacks: connection lost, message arrived, delivery complete */
    MQTTClient_setCallbacks(client, NULL, connectionLost, messageArrived, deliveryComplete);

    /* 3) Connect to the broker */
    conn_opts.keepAliveInterval = 20;
    conn_opts.cleansession = 1;
    rc = MQTTClient_connect(client, &conn_opts);
    if (rc != MQTTCLIENT_SUCCESS)
    {
        printf("Failed to connect, return code %d\n", rc);
        return EXIT_FAILURE;
    }
    printf("[ASYNC] Connected to broker at %s\n", ADDRESS);

    /* 4) Subscribe to a topic */
    rc = MQTTClient_subscribe(client, TOPIC, QOS);
    if (rc != MQTTCLIENT_SUCCESS)
    {
        printf("Failed to subscribe, return code %d\n", rc);
        return EXIT_FAILURE;
    }
    printf("[ASYNC] Subscribed to topic: %s\n", TOPIC);

    /* 5) Let the main thread continue. The library handles incoming messages
          in a background thread and invokes the callbacks automatically.
          For demo, we just wait until the user presses Enter to exit. */
    printf("[ASYNC] Press Enter to quit...\n");
    getchar();

    /* 6) Disconnect and cleanup */
    MQTTClient_disconnect(client, 10000);
    MQTTClient_destroy(&client);

    return EXIT_SUCCESS;
}